### PR TITLE
[Enhancement] pushdown predicates after mv union rewrite

### LINKED ISSUE #23004: predicates can be pushdown after join mv union rewrite
## Enhancement

now, for join queries, after mv union rewrite, precates may be added directly into join node, which can be further pushdown.

for mv:
create materialized view join_agg_union_mv_1" +
                " distributed by hash(v1)" +
                " as " +
                " SELECT t02.v1 as v1, test_all_type2.t1d," +
                " sum(test_all_type2.t1c) as total_sum, count(test_all_type2.t1c) as total_num" +
                " from t02 join test_all_type2" +
                " on t02.v1 = test_all_type2.t1d" +
                " where t02.v1 < 100" +
                " group by v1, test_all_type2.t1d

for query:
" SELECT t02.v1 as v1, test_all_type2.t1d," +
                " sum(test_all_type2.t1c) as total_sum, count(test_all_type2.t1c) as total_num" +
                " from t02 join test_all_type2" +
                " on t02.v1 = test_all_type2.t1d" +
                " where t02.v1 < 120" +
                " group by v1, test_all_type2.t1d"

PLAN FRAGMENT 0
 OUTPUT EXPRS:1: v1 | 7: t1d | 14: sum | 15: count
  PARTITION: UNPARTITIONED

  RESULT SINK

  14:EXCHANGE

PLAN FRAGMENT 1
 OUTPUT EXPRS:
  PARTITION: HASH_PARTITIONED: 39: v1, 41: t1d

  STREAM DATA SINK
    EXCHANGE ID: 14
    UNPARTITIONED

  13:Project
  |  <slot 1> : 39: v1
  |  <slot 7> : 41: t1d
  |  <slot 14> : 43: sum
  |  <slot 15> : 44: count
  |  
  12:AGGREGATE (merge finalize)
  |  output: sum(43: sum), sum(44: count)
  |  group by: 39: v1, 41: t1d
  |  
  11:EXCHANGE

PLAN FRAGMENT 2
 OUTPUT EXPRS:
  PARTITION: RANDOM

  STREAM DATA SINK
    EXCHANGE ID: 11
    HASH_PARTITIONED: 39: v1, 41: t1d

  10:AGGREGATE (update serialize)
  |  STREAMING
  |  output: sum(40: sum), sum(42: count)
  |  group by: 39: v1, 41: t1d
  |  
  0:UNION
  |  
  |----9:EXCHANGE
  |    
  6:EXCHANGE

PLAN FRAGMENT 3
 OUTPUT EXPRS:
  PARTITION: RANDOM

  STREAM DATA SINK
    EXCHANGE ID: 09
    RANDOM

  8:Project
  |  <slot 35> : 16: v1
  |  <slot 36> : 18: total_sum
  |  <slot 37> : 16: v1
  |  <slot 38> : 19: total_num
  |  
  7:OlapScanNode
     TABLE: join_agg_union_mv_1
     PREAGGREGATION: ON
     partitions=1/1
     rollup: join_agg_union_mv_1
     tabletRatio=6/6
     tabletList=10590,10592,10594,10596,10598,10600
     cardinality=1
     avgRowSize=7.0
     numNodes=0

PLAN FRAGMENT 4
 OUTPUT EXPRS:
  PARTITION: RANDOM

  STREAM DATA SINK
    EXCHANGE ID: 06
    RANDOM

  5:AGGREGATE (update finalize)
  |  output: sum(23: t1c), count(23: t1c)
  |  group by: 20: v1, 24: t1d
  |  
  4:HASH JOIN
  |  join op: INNER JOIN (BUCKET_SHUFFLE)
  |  colocate: false, reason: 
  |  equal join conjunct: 20: v1 = 24: t1d
  **|  other predicates: 20: v1 < 120, 20: v1 > 99**
  |  
  |----3:EXCHANGE
  |    
  1:OlapScanNode
     TABLE: t02
     PREAGGREGATION: ON
     PREDICATES: 20: v1 < 120
     partitions=1/1
     rollup: t02
     tabletRatio=3/3
     tabletList=10564,10566,10568
     cardinality=500000
     avgRowSize=1.0
     numNodes=0

PLAN FRAGMENT 5
 OUTPUT EXPRS:
  PARTITION: RANDOM

  STREAM DATA SINK
    EXCHANGE ID: 03
    BUCKET_SHUFFLE_HASH_PARTITIONED: 24: t1d

  2:OlapScanNode
     TABLE: test_all_type2
     PREAGGREGATION: ON
     PREDICATES: 24: t1d < 120
     partitions=1/1
     rollup: test_all_type2
     tabletRatio=3/3
     tabletList=10555,10557,10559
     cardinality=500000
     avgRowSize=2.0
     numNodes=0

### PR DESCRIPTION
**Pull request by ABingHuang**
## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes:

Fixes #23004

## Problem Summary(Required):

pushdown predicates after mv union rewrite
## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 371 / 379 (97.89%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|34|37|91.89%|[813, 822, 1051]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/JoinPredicatePushdown.java|331|336|98.51%|[180, 189, 199, 200, 201]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownPredicateJoinRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownJoinOnClauseRule.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 371 / 379 (97.89%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|34|37|91.89%|[813, 822, 1051]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/JoinPredicatePushdown.java|331|336|98.51%|[180, 189, 199, 200, 201]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownPredicateJoinRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownJoinOnClauseRule.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 34 / 37 (91.89%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|34|37|91.89%|[813, 822, 1051]|

**Review by murphyatwork — APPROVED** [review]

**Review by murphyatwork — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java [hunk 1]
```diff
@@ -1013,18 +1028,49 @@ protected OptExpression queryBasedRewrite(RewriteContext rewriteContext, ScalarO
...
 
+    // pushdown predicates on join nodes
+    // the OptExpression will be modified in place
+    private OptExpression pushdownPredicatesForJoin(OptExpression optExpression, ScalarOperator predicate) {
+        if (!(optExpression.getOp() instanceof LogicalJoinOperator)) {
+            if (predicate != null) {
+                // predicate can not be pushdown, we should add it it optExpression
+                Operator.Builder builder = OperatorBuilderFactory.build(optExpression.getOp());
+                builder.withOperator(optExpression.getOp());
+                builder.setPredicate(predicate);
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:1047]: May lost the origin predicate in this operator

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java [hunk 2]
```diff
@@ -1013,18 +1028,49 @@ protected OptExpression queryBasedRewrite(RewriteContext rewriteContext, ScalarO
...
+                Operator newQueryOp = builder.build();
+                return OptExpression.create(newQueryOp, optExpression.getInputs());
+            } else {
+                return optExpression;
+            }
+        }
+        OptExpression newJoin = doPushdownPredicate(optExpression, predicate);
+        // pushdown predicates in children
+        List<OptExpression> children = Lists.newArrayList();
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:1056]: If there are other operator between join and scan, could this ensure psuh predicate to scan?

Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:1056]: In the mv rewrite case, there are only join and scan operator.

**Review by Youngwb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 36 / 38 (94.74%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|36|38|94.74%|[815, 1053]|

**Review by murphyatwork — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 36 / 38 (94.74%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|36|38|94.74%|[815, 1053]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 36 / 38 (94.74%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|36|38|94.74%|[815, 1053]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 36 / 38 (94.74%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|36|38|94.74%|[815, 1053]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 36 / 38 (94.74%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|36|38|94.74%|[815, 1053]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 34 / 37 (91.89%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|34|37|91.89%|[815, 824, 1053]|

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**